### PR TITLE
BIM: use external defining sketch geometry

### LINKED ISSUE #28239: BIM: External defining geometry in a sketch doesn't generate Walls
### Problem description

As per title, creating a BIM Wall out of a sketch will only generate walls along defining geometry you created yourself, if you have any defining imported geometry, it will be ignored.

Example file, remove `.zip` from the filename: [external defining geometry no BIM wall 1.2.dev.FCStd.zip](https://github.com/user-attachments/files/25872166/external.defining.geometry.no.BIM.wall.1.2.dev.FCStd.zip)

### Workbench affected?

BIM

### Steps to reproduce

- create a source sketch / shape to have something to import 
- create a destination sketch, import some geometry into it and make sure it's set as defining, not as construction, requires 1.1+ for this
- optionally, add some defining geometry of your own
- select this destination sketch and click the button to create a BIM Wall out of it

### Expected behavior

- all defining geometry, regardless of being your own or imported, should generate walls

### Actual behavior

- if the sketch only contains defining external geometry, no walls get generated and you get a `Wall has a null shape` error
- if the sketch contains any defining geometry you created yourself, only that geometry will be used to create walls, all imported defining geometry will be ignored

sketch + result:

Image

Image

### Development version About Info (in Safe Mode)

```shell
OS: Windows 10 build 19045
Architecture: x86_64
Version: 1.2.0dev.20260303 (Git shallow)
Build date: 2026/03/03 22:57:35
Build type: Release
Branch: grafted,grafted
Hash: 637463ac354f093e14dacfc5059eb70c2b1216fd
Python 3.11.14, Qt 6.8.3, Coin 4.0.3, Vtk 9.3.1, boost 1_86, Eigen3 3.4.0, PySide 6.8.3
shiboken 6.8.3, xerces-c 3.3.0, IfcOpenShell 0.8.2, OCC 7.8.1
Locale: English/United States (en_US)
Navigation Style/Orbit Style/Rotation Mode: Blender/Rounded Arcball/Window center
Stylesheet/Theme/QtStyle: FreeCAD.qss/FreeCAD Dark/
Logical DPI/Physical DPI/Pixel Ratio: 96/81.5973/1
```

### Last known good version (optional)

no version seems to support this (1.0 doesn't even apply as there you can't have external defining geometry in a sketch)

### PR DESCRIPTION
**Pull request by yangyu0330**
## Summary
- collect external defining sketch geometry alongside internal non-construction geometry for BIM sketch bases
- reuse the shared sketch edge collector in Wall, CurtainWall, and Structure generation paths
- add a Wall regression test for a sketch whose only wall baseline is an external defining edge

## Tests
- python -m py_compile src\Mod\BIM\ArchSketchObject.py src\Mod\BIM\ArchWall.py src\Mod\BIM\ArchCurtainWall.py src\Mod\BIM\ArchStructure.py src\Mod\BIM\bimtests\TestArchWall.py
- git diff --check
- uvx pre-commit run --files src\Mod\BIM\ArchSketchObject.py src\Mod\BIM\ArchWall.py src\Mod\BIM\ArchCurtainWall.py src\Mod\BIM\ArchStructure.py src\Mod\BIM\bimtests\TestArchWall.py

Fixes #28239

### CODE REVIEW

**Comment by chennes** [comment]
Hi! Due to huge spike of PRs submitted to the project mostly done solely using AI tooling we will be more strict with our current policy about the submissions.

Since we are not fully sure how much human expertise and knowledge was put in making of this PR and it seems that it was AI generated without much review or human action we are closing it.

If you feel that we made mistake and closing your contribution was unjustified - feel free to contact us via any means you feel. No decision is final and if we misjudged something - we are happy to revise the choice.

Thanks for understanding!

**Comment by Roy-043** [comment]
The code in the PR is good. I am therefore reopening it.

**Comment by Roy-043** [comment]
@yangyu0330 Sorry for the premature closing of this PR. I was not aware of it until yesterday. I have reopened it now.

@paullee0 AFAICT the proposed code works fine and solves the issue. But can you please check as well? Thanks.

Test file (remove .zip, open and recompute):
[sketch_external.FCStd.zip](https://github.com/user-attachments/files/31636584/sketch_external.FCStd.zip)

File: src/Mod/BIM/ArchSketchObject.py [hunk 1]
```diff
@@ -25,6 +25,49 @@
...
 
+def getSketchDefiningEdges(sketch, selected_edges=None, supported_geometry=None):
+    import Part
+    import Sketcher
+
+    if supported_geometry is None:
+        supported_geometry = (Part.LineSegment, Part.Circle, Part.ArcOfCircle, Part.Ellipse)
+
+    selected_edges = {str(edge) for edge in selected_edges or []}
+    edges = []
```
Comment (Roy-043) [inline src/Mod/BIM/ArchSketchObject.py:36]:
```suggestion
def getSketchDefiningEdges(sketch, selected_edges=[], supported_geometry=None):
    import Part
    import Sketcher

    if supported_geometry is None:
        supported_geometry = (Part.LineSegment, Part.Circle, Part.ArcOfCircle, Part.Ellipse)

    edges = []
```
There is no need to convert `selected_edges` items to string. Also no need to create a set from the list.

**Review by Roy-043 — COMMENTED** [review]
Minor issue.

**Comment by paullee0** [comment]
Thanks @Roy-043 for the heads-up.

Just had some comments on FC forum on related issue quoted below, would have another look noting a PR is submitted though :D

https://forum.freecad.org/viewtopic.php?p=903957#p903957 : -
"Not sure if external geometry is still subject to TNP, particularly referencing other sketches. This is first factor to consider supporting that.

ArchSketch already had experimental feature which 'copy carbon' another ArchSketch edges/wall segment, and it should support the individual Width, Align, Offset to my memory, and it is 'TNP-tolerant (feature not exposed though for reasons). The 'rich-information' in the former and TNP-tolerant in the latter IMHO are some basic factors to consider what 'external geometries' should be supported and developed.


What is the workflow with 'external geometries' you think good for architectural modelling and development? :)"

**Comment by paullee0** [comment]
OK, thanks, had a quick study :)

Need to get familiar with a few things first e.g. Sketcher.ExternalGeometryFacade(geometry), facade.testFlag("Defining") etc. - any pointers?

What are the indexes for those External Geometries?

Could (/Should) External Geometries be limited to reference to another Sketch (ArchSketch) instead of every other object has Shape?

Then, algorithm to sort those included edges for their corresponding Width, Align, Offset etc. needs to be tweaked.

https://github.com/yangyu0330/FreeCAD/blob/3c11e006e3e4cb82a6d839eea2430cce4e5aad57/src/Mod/BIM/ArchWall.py#L882-L884

                        widths = ArchSketchObject.sortSketchWidth(
                            obj.Base, obj.OverrideWidth, obj.ArchSketchEdges
                        )

**Comment by Roy-043** [comment]
### Code explanation

The external geometry of a sketch is stored in its `ExternalGeo` property. The first 2 elements in that list are the axes of the sketch, then come the edges/vertices referenced from other objects.

To get the 1st external edge of a sketch object:
```python
geometry = obj.ExternalGeo[2:][0]
```

That edge seems completely separate from the sketch, but contains some data that references the sketch:
```python
geometry.getExtensions()
# which may return:
# [<SketchGeometryExtension ("1") >, <ExternalGeometryExtension ("Sketch.;g1;SKT",{Defining}) >]
```

With that data `Sketcher.ExternalGeometryFacade()` can create an `ExternalGeometryFacadePy` object:
```
facade = Sketcher.ExternalGeometryFacade(geometry)
```
You can then easily access the 'extension' data:
```
facade.Ref
# which may return:
# 'Sketch.;g1;SKT'

facade.testFlag("Defining")
# which may return:
# True
```

### Indices

The code in this PR ensures the indices for the external geometry are -3, -4 etc. So the same as you would see in the GUI.

### Which defining external geometry to consider?

I would say all. But I do not know the ArchSketch algorithms.

File: src/Mod/BIM/ArchSketchObject.py [hunk 2]
```diff
@@ -25,6 +25,49 @@
...
+        if selected_edges:
+            if str(index) not in selected_edges:
+                continue
+        elif facade.Construction:
+            continue
+
+        if isinstance(facade.Geometry, supported_geometry):
+            edges.append(facade.Geometry.toShape())
+
+    external_geometry = list(getattr(sketch, "ExternalGeo", []))
```
Comment (Roy-043) [inline src/Mod/BIM/ArchSketchObject.py]:
```suggestion
    external_geometry = sketch.ExternalGeo
```
Another small simplification.

**Review by Roy-043 — CHANGES_REQUESTED** [review]

**Review by Roy-043 — APPROVED** [review]

**Comment by Roy-043** [comment]
I have taken the liberty of applying two minor tweaks.

**Comment by paullee0** [comment]
> I have taken the liberty of applying two minor tweaks.

Checking the code again ;)
